feat(signals): add `signalMethod` (#4597)
ngrx/platform: projects/ngrx.io/content/guide/signals/signal-method.md

export class NumbersService {
  readonly logDoubledNumber = signalMethod<number>((num) => {
    const double = num * 2;
    console.log(double);
  });
}

@Component({ /* ... */ })
export class NumbersComponent implements OnInit {
  readonly numbersService = inject(NumbersService);

  ngOnInit(): void {
    const value = signal(2);
    // 👇 Uses the injection context of the `NumbersService`, which is root.
    this.numbersService.logDoubledNumber(value);
  }
}
```

Here, the `effect` outlives the component, which would produce a memory leak.

## Manual Cleanup

When a `signalMethod` is created in an ancestor injection context, it's necessary to explicitly provide the caller injector to ensure proper cleanup:

```ts
@Component({ /* ... */ })
export class NumbersComponent implements OnInit {
  readonly numbersService = inject(NumbersService);
  readonly injector = inject(Injector);

  ngOnInit(): void {
    const value = signal(1);
    // 👇 Providing the `NumbersComponent` injector
    // to ensure cleanup on component destroy.
    this.numbersService.logDoubledNumber(value, {
      injector: this.injector,
    });
  
    // 👇 No need to provide an injector for static values.

>     this.logDoubledNumber(2);

Reviewer: ```suggestion
    this.numbersService.logDoubledNumber(2);
```

Careful, when `signalMethod` is used in a service which is provided in `root`:

```ts

@Injectable({providedIn: 'root'})
export class NumbersService {
  readonly logDoubledNumber = signalMethod<number>(num => {
    const double = num * 2;
    console.log(double);
  })
}

@Component({ /* ... */})
export class NumbersComponent {
  readonly logDoubledNumber = inject(NumbersService).logDoubledNumber;

  ngOnInit(): void {
    this.logDoubledNumber(1);

    const value = signal(2);
    // 👇 uses the injection context of the `NumbersService`, which is root.
    this.logDoubledNumber(value);
  }
}
```

Here, the `effect` outlives the component, which would produce a memory leak.

As a consequence, try to call the "processor function" always in an injection context:

```ts

@Component({ /* ... */})
export class NumbersComponent {
  readonly logDoubledNumber = inject(NumbersService).logDoubledNumber;

  ngOnInit(): void {
    this.logDoubledNumber(1);

    const value = signal(2);
    // 👇 uses the injection context of the `NumbersService`, which is root.
    this.logDoubledNumber(value);
  }
}
```

> ## Providing an Injector

Reviewer 2: ```suggestion
## Manual Cleanup
```

> If you cannot run the "processor function" in an injection context, you can also provide an injector manually:

Reviewer 2: for consistency with other pages from `@ngrx/signals` docs, impersonal style is used:

```suggestion
When a `signalMethod` is created in an ancestor injection context, it's necessary to explicitly provide the caller injector to ensure proper cleanup:
```

```ts

> 
> @Component({ /* ... */})
> export class NumbersComponent {
>   readonly logDoubledNumber = inject(NumbersService).logDoubledNumber;
>   readonly injector = inject(Injector);
> 
>   ngOnInit(): void {
>     this.logDoubledNumber(1);
> 
>     const value = signal(2);
>     // 👇 uses the injection context of the `NumbersService`, which is root.
>     this.logDoubledNumber(value, {injector: this.injector});
>   }
> }

Reviewer 2: ```suggestion
@Component({ /* ... */ })
export class NumbersComponent implements OnInit {
  readonly numbersService = inject(NumbersService);
  readonly injector = inject(Injector);

  ngOnInit(): void {
    const value = signal(1);
    // 👇 Providing the `NumbersComponent` injector
    // to ensure cleanup on component destroy.
    this.numbersService.logDoubledNumber(value, {
      injector: this.injector,
    });
  
    // 👇 No need to provide an injector for static values.
    this.logDoubledNumber(2);
  }
}
```

```

> Whereas the "processor function" doesn't require an active injection context, the call of `signalMethod` does:
> 
> ```ts
> 
> @Component({ /* ... */})
> export class NumbersComponent {
> // 👇 Would cause a runtime error
>   ngOnInit() {
>     const logDoubledNumber = signalMethod<number>(num => console.log(num * 2));
>   }
> }
> ```
> 
> In these cases, you also have to provide the injector manually:
> 
> ```ts
> 
> @Component({ /* ... */})
> export class NumbersComponent {
>   readonly injector = inject(Injector);
> 
>   ngOnInit() {
>     // 👇 Works now
>     const logDoubledNumber = signalMethod<number>(num => console.log(num * 2), {injector: this.injector});
>   }
> }
> ```

Reviewer 2: This can be a separate section:

```suggestion
## Initialization Outside of Injection Context

The `signalMethod` must be initialized within an injection context. To initialize it outside an injection context, it's necessary to provide an injector as the second argument:

```ts
@Component({ /* ... */ })
export class NumbersComponent implements OnInit {
  readonly injector = inject(Injector);

  ngOnInit() {
    const logDoubledNumber = signalMethod<number>(
      (num) => console.log(num * 2),
      { injector: this.injector },
    );
  }
}
```
```

> ## Advantages over simple effect

Reviewer 2: I like this section! We should just update the heading to the title case format for consistency:

```suggestion
## Advantages over Effect
```

At first sight, `signalMethod`, might be the same as `effect`:

```ts

> 
> @Component({ /* ... */})
> export class NumbersComponent {
>   readonly value = signal(2);
>   readonly logDoubledNumberEffect = effect(() => {
>     const double = num * 2;
>     console.log(double);
>   })
> 
>   constructor(): void {
>     this.logDoubledNumber(value);
>   }
> }

Reviewer 2: - effect implementation is invalid
- signalMethod is missing

```suggestion
@Component({ /* ... */ })
export class NumbersComponent {
  readonly num = signal(2);
  readonly logDoubledNumberEffect = effect(() => {
    console.log(this.num() * 2);
  });
  readonly logDoubledNumber = signalMethod<number>((num) => {
    console.log(num * 2);
  });

  constructor() {
    this.logDoubledNumber(this.num);
  }
}
```

```

However, `signalMethod` offers three distinctive advantages over `effect`:

> - **Flexible Parameter**: The parameter can also be a simple number, not just a Signal.

Reviewer 2: ```suggestion
- **Flexible Input**: The input argument can be a static value, not just a signal. Additionally, the processor function can be called multiple times with different inputs.
```

  }
}
```

In these cases, you also have to provide the injector manually:

```ts

@Component({ /* ... */})
export class NumbersComponent {
  readonly injector = inject(Injector);

  ngOnInit() {
    // 👇 Works now
    const logDoubledNumber = signalMethod<number>(num => console.log(num * 2), {injector: this.injector});
  }
}
```

## Advantages over simple effect

On first sight, `signalMethod`, might be the same as `effect`:

```ts

@Component({ /* ... */})
export class NumbersComponent {
  readonly value = signal(2);
  readonly logDoubledNumberEffect = effect(() => {
    const double = num * 2;
    console.log(double);
  })

  constructor(): void {
    this.logDoubledNumber(value);
  }
}
```

However, `signalMethod` offers three distinctive advantages over `effect`:

- **Flexible Parameter**: The parameter can also be a simple number, not just a Signal.
- **No Injection Context Required**: Unlike an `effect`, which requires an injection context or an Injector, `signalMethod`'s "processor function" can be called without an injection context.
- **Explicit Tracking**: Only the Signal of the parameter is tracked, while Signals within the "processor function" stay untracked.

> <div class="alert is-helpful">

Author: I've now replaced the alert with a new section. Let me know what you think.

> <div class="alert is-helpful">

Reviewer: It looks good to me, thanks @rainerhahnekamp !
I'll go over it once again when Marko's comments are resolved.
